### PR TITLE
chore(frontend): improve response types for Coingecko requests

### PR DESCRIPTION
**Pull request by AntonioVentilii-DFINITY**
# Motivation

After PR #1737 , the response type of function `exchangeRateERC20ToUsd` returns the prop `usd_market_cap` as `number` (and not possibly `undefined`). To be type-safe, we create a new type specific to this function.

# Changes

- Create type `CoingeckoSimpleTokenPriceResponse` that remove optionality of `usd_market_cap`.
- Use the type for function `exchangeRateERC20ToUsd`.
- Adapt function that fetches prices to have variable type on response.

### CODE REVIEW

**Comment by AntonioVentilii-DFINITY** [comment]
@peterpeterparker 

I actualy have done a first version a bit more complex that relies directly on the value of input parameters for the props of the output.
For example: if `include_market_cap` is `true`, then `usd_market_cap` is not `undefined`; if `include_24hr_vol` is `true`, then `usd_24h_vol` is not `undefined`; etc etc.

You can see the commit here: [
commit d9d836ffbbc1714a04af76c8659657d87cd44df2](https://github.com/dfinity/oisy-wallet/pull/1738/commits/d9d836ffbbc1714a04af76c8659657d87cd44df2)

I think it is overkill, but theoretically it is the consistent way of doing this. Let me know if you prefer that commit or this one

File: src/frontend/src/lib/types/coingecko.ts [hunk 1]
```diff
@@ -53,3 +53,12 @@ export interface CoingeckoSimplePrice {
 }
 
 export type CoingeckoSimplePriceResponse = Record<CoingeckoCoinsId | string, CoingeckoSimplePrice>;
+
+export interface CoingeckoSimpleTokenPrice extends CoingeckoSimplePrice {
```
Comment (peterpeterparker) [inline src/frontend/src/lib/types/coingecko.ts:57]: isn't `Omit<CoingeckoSimplePrice, 'usd_market_cap'>`

or better

```
export type CoingeckoSimpleTokenPrice = Omit<CoingeckoSimplePrice, 'usd_market_cap'> & Required<Pick<CoingeckoSimplePrice, 'usd_market_cap'>
```

File: src/frontend/src/lib/types/coingecko.ts [hunk 2]
```diff
@@ -53,3 +53,12 @@ export interface CoingeckoSimplePrice {
...
+	usd_market_cap: number;
+}
+
+export type CoingeckoSimpleTokenPriceResponse = Record<
```
Comment (peterpeterparker) [inline src/frontend/src/lib/types/coingecko.ts:61]: Just an idea. What do you think?

```
export type CoingeckoResponse<T> = Record<CoingeckoCoinsId | string, T>;
```

which you can use in both `CoingeckoSimplePriceResponse` and `CoingeckoSimpleTokenPriceResponse`

```
export type CoingeckoSimpleTokenPriceResponse = CoingeckoResponse<CoingeckoSimpleTokenPrice>
```

that way you can declare `fetchCoingecko` as returning a type extending response

```
const fetchCoingecko = async <T extends >({
  ...
  }): Promise<CoingeckoResponse<T> | null> => {
```

and modify caller to pass the inner type for T

```
export const simpleTokenPrice = async ({
	id,
	contract_addresses,
	...rest
}: CoingeckoSimpleTokenPriceParams): Promise<CoingeckoSimpleTokenPriceResponse | null> =>
	fetchCoingecko<CoingeckoSimpleTokenPrice>({
```

The advantage is that `fetchCoingecko` returns become a bit more descriptive.

File: src/frontend/src/lib/rest/goincecko.rest.ts [hunk 3]
```diff
@@ -50,13 +51,13 @@ export const simpleTokenPrice = async ({
 		])
 	});
 
-const fetchCoingecko = async ({
+const fetchCoingecko = async <T>({
```
Comment (peterpeterparker) [inline src/frontend/src/lib/rest/goincecko.rest.ts:54]: Maybe?

```
const fetchCoingecko = async <T extends CoingeckoSimplePrice | CoingeckoSimpleTokenPrice>({
```

File: src/frontend/src/lib/types/coingecko.ts (hunk 2 above)
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/types/coingecko.ts:61]: agreed! one doubt: why would `fetchCoingecko` have a type extending? can it be the same if:

```typescript
const fetchCoingecko = async <T>({
	endpointPath,
	queryParams
}: {
	endpointPath: string;
	queryParams: string;
}): Promise<CoingeckoResponse<T> | null> => {
...
```

Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/types/coingecko.ts:61]: ah! i saw your next comment!

File: src/frontend/src/lib/rest/goincecko.rest.ts (hunk 3 above)
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/rest/goincecko.rest.ts:54]: yes, more descriptive indeed!

**Review by peterpeterparker — APPROVED** [review]
LGTM, thx

**Pull request merged**